File: Day20/delete-duplicate-folders-in-system.py
class Trie:
    #serial: str = ''
    #children: dict

    def __init__(self):
        self.serial = ''
        self.children = dict()

class Solution:
    def deleteDuplicateFolder(self, paths: List[List[str]]) -> List[List[str]]:
        root = Trie()

        for path in paths:#(n*size(x))
            currNode = root
            for node in path:
                if node not in currNode.children:
                    currNode.children[node] = Trie()
                currNode = currNode.children[node]
        
        freq = Counter()

        def construct(node: Trie) -> None: #(n*size(x) => x is each folder length)
            if not node.children:
                return
            
            v = []

            for folder, child in node.children.items():
                construct(child)
                v.append(folder+'(' + child.serial + ')')
            
            v.sort()
            node.serial = "".join(v)
            freq[node.serial]+=1
        
        construct(root)

        nonDuplicateFolders= []
        path = []

        def operate(node: Trie) -> None: #(n*size(x) => x is each folder length)
            if freq[node.serial] > 1:
                return
            if path:
                nonDuplicateFolders.append(path[:])
            
            for folder, child in node.children.items():
                path.append(folder)
                operate(child)
                path.pop()

        operate(root)
        return nonDuplicateFolders

        

File: Day20/delete-duplicate-folders-in-system.js
class Trie {
    constructor() {
        this.serial = "";
        this.children = {};
    }
}

var deleteDuplicateFolder = function(paths) {
    const root = new Trie();

    for (let path of paths) {
        let node = root;
        for (let folder of path) {
            if (!node.children[folder]) {
                node.children[folder] = new Trie();
            }
            node = node.children[folder];
        }
    }

    const freq = new Map();

    function construct(node) {
        if (Object.keys(node.children).length === 0) return;

        const v = [];
        for (let [folder, child] of Object.entries(node.children)) {
            construct(child);
            v.push(folder + '(' + child.serial + ')');
        }
        v.sort();
        node.serial = v.join('');
        freq.set(node.serial, (freq.get(node.serial) || 0) + 1);
    }

    construct(root);

    const result = [];
    const path = [];

    function dfs(node) {
        if (freq.get(node.serial) > 1) return;
        if (path.length > 0) result.push([...path]);

        for (let [folder, child] of Object.entries(node.children)) {
            path.push(folder);
            dfs(child);
            path.pop();
        }
    }

    dfs(root);
    return result;
};
